File: Aoc.Fsharp/Solutions/Day02.fs
module Aoc.Solutions.Day02

open Aoc.Runner

type Command = { dir: string; dist: int }
type Position = { depth: int; horz: int; aim: int}

let parse (line: string) =
    let parts = line.Split()
    { dir = parts[0]; dist = int parts[1] }

let parseInput (input: string) =
    input.Split('\n')
        |> Seq.map parse

let executeOne pos command =
    match command with
    | { dir = "forward"; dist = d  } -> { pos with horz = pos.horz + d }
    | { dir = "up"; dist = d  } -> { pos with depth = pos.depth - d }
    | { dir = "down"; dist = d  } -> { pos with depth = pos.depth + d }
    | _ -> failwith("Not implemented")


let executeAim pos command =
    match command with
    | { dir = "forward"; dist = d  } ->
        { pos with horz = pos.horz + d; depth = pos.depth + d * pos.aim }
    | { dir = "up"; dist = d  } -> { pos with aim = pos.aim - d }
    | { dir = "down"; dist = d  } -> { pos with aim = pos.aim + d }
    | _ -> failwith("Not implemented")


let execute fn inputs =
    inputs
    |> Seq.fold fn { depth = 0; horz = 0 ; aim = 0}

type Day02() =
    inherit Day()

    override _.SolveA(input) =
        let position =
            parseInput input
            |> execute executeOne
        string (position.depth * position.horz)

    override _.SolveB(input) =
        let position =
            parseInput input
            |> execute executeAim
        string (position.depth * position.horz)

    override this.Tests =
        [ new Test(
              name = "basic part A",
              input =
                  @"forward 5
down 5
forward 8
up 3
down 8
forward 2",
              expectedOutput = "150",
              solve = fun x -> this.SolveA(x)
         );
        new Test(
              name = "part B",
              input =
                  @"forward 5
down 5
forward 8
up 3
down 8
forward 2",
              expectedOutput = "900",
              solve = fun x -> this.SolveB(x)
          ) ]
        |> ResizeArray

File: Aoc.Csharp/Solutions/Day02.cs
using SharpResult.FunctionalExtensions;

namespace Aoc.Solutions;
public class Day02 : Day
{
    record Pos(int H, int D, int A);
    record Command(string C, int X)
    {
        public static Command FromString(string s)
        {
            var parts = s.Split();
            return new(parts[0], int.Parse(parts[1]));
        }
    }

    public override string SolveA(string input)
    {
        var pos = new Pos(0, 0, 0);
        foreach (var comm in input.Split('\n').Select(Command.FromString))
        {
            pos = comm switch
            {
                ("forward", var x) => pos with { H = pos.H + x },
                ("up", var x) => pos with { D = pos.D - x },
                ("down", var x) => pos with { D = pos.D + x },
                _ => throw new Exception()
            };
        }
        return (pos.H * pos.D).ToString();
    }

    public override string SolveB(string input)
    {
        var pos = new Pos(0, 0, 0);
        foreach (var comm in input.Split('\n').Select(Command.FromString))
        {
            pos = comm switch
            {
                ("forward", var x) => pos with { H = pos.H + x, D = pos.D + x * pos.A },
                ("up", var x) => pos with { A = pos.A - x },
                ("down", var x) => pos with { A = pos.A + x },
                _ => throw new Exception()
            };
        }
        return (pos.H * pos.D).ToString();
    }

    public Day02()
    {
        Tests = new()
        {
            new("A", @"forward 5
down 5
forward 8
up 3
down 8
forward 2", "150", SolveA)
        };
    }
}